tests/fipmapper/test_disjointsets.py

```diff
@@ -0,0 +1,140 @@
+            [
+                {"REGION": "A", "ZONE": "L", "FIPNUM": 2, "ROOT": ("A", "L", 2)},
+                {"REGION": "A", "ZONE": "U", "FIPNUM": 1, "ROOT": ("A", "U", 1)},
+            ],
+        ),
+        (
+            # Two zones, one FIPNUM, gives one group in return:
+            {"region2fipnum": {"A": 1}, "zone2fipnum": {"U": [1], "L": [1]}},
+            [
+                {"REGION": "A", "ZONE": "L", "FIPNUM": 1, "ROOT": ("A", "U", 1)},
+                {"REGION": "A", "ZONE": "U", "FIPNUM": 1, "ROOT": ("A", "U", 1)},
+            ],
+        ),
+        (
+            # Two regions, one FIPNUM, gives one group in return:
+            {"region2fipnum": {"A": 1, "B": 1}, "zone2fipnum": {"U": 1}},
+            [
+                {"REGION": "A", "ZONE": "U", "FIPNUM": 1, "ROOT": ("B", "U", 1)},
+                {"REGION": "B", "ZONE": "U", "FIPNUM": 1, "ROOT": ("B", "U", 1)},
+            ],
+        ),
+        (
+            # Two regions, two zones, 4 FIPNUMs -> 4 groups:
+            {
+                "region2fipnum": {"A": [2, 1], "B": [3, 4]},
+                "zone2fipnum": {"U": [1, 3], "L": [2, 4]},
+            },
+            [
+                {"REGION": "A", "ZONE": "L", "FIPNUM": 2, "ROOT": ("A", "L", 2)},
+                {"REGION": "A", "ZONE": "U", "FIPNUM": 1, "ROOT": ("A", "U", 1)},
+                {"REGION": "B", "ZONE": "L", "FIPNUM": 4, "ROOT": ("B", "L", 4)},
+                {"REGION": "B", "ZONE": "U", "FIPNUM": 3, "ROOT": ("B", "U", 3)},
+            ],
+        ),
+        (
+            # Two regions, two zones, 2 horizontal FIPNUMs -> 2 groups:
+            {
+                "region2fipnum": {"A": [2, 1], "B": [1, 2]},
+                "zone2fipnum": {"U": 1, "L": [2, 2]},  # Double for L is ignored.
+            },
+            [
+                {"REGION": "A", "ZONE": "L", "FIPNUM": 2, "ROOT": ("B", "L", 2)},
+                {"REGION": "A", "ZONE": "U", "FIPNUM": 1, "ROOT": ("B", "U", 1)},
+                {"REGION": "B", "ZONE": "L", "FIPNUM": 2, "ROOT": ("B", "L", 2)},
+                {"REGION": "B", "ZONE": "U", "FIPNUM": 1, "ROOT": ("B", "U", 1)},
+            ],
+        ),
+        (
+            # Two regions, two zones, 2 region-wise FIPNUMs -> 2 groups:
+            {
+                "region2fipnum": {"A": 1, "B": 2},
+                "zone2fipnum": {"U": [1, 2], "L": [1, 2]},
+            },
+            [
+                {"REGION": "A", "ZONE": "L", "FIPNUM": 1, "ROOT": ("A", "U", 1)},
+                {"REGION": "A", "ZONE": "U", "FIPNUM": 1, "ROOT": ("A", "U", 1)},
+                {"REGION": "B", "ZONE": "L", "FIPNUM": 2, "ROOT": ("B", "U", 2)},
+                {"REGION": "B", "ZONE": "U", "FIPNUM": 2, "ROOT": ("B", "U", 2)},
+            ],
+        ),
+        (
+            # Almost three-by-three reservoir, used in demonstration
+            # of the problem solved, involving all "cruxes":
+            {
```

Comment:
I need a bit more context to be able to review this. In particular, I tried to understand the tests by inspecting them and the documentation, but did not fully get it. Could you provide a more in-details explanation of the input and output of this example? :)

Reply:
ASCII-art added, and the example is elevated to be the first test. Hopefully it helps..